receiver/otlpreceiver/internal/arrow/arrow.go

```diff
@@ -0,0 +1,119 @@
+package arrow
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/obsreport"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	arrowpb "github.com/lquerel/otel-arrow-adapter/api/collector/arrow/v1"
+	batchEvent "github.com/lquerel/otel-arrow-adapter/pkg/otel/batch_event"
+	"github.com/lquerel/otel-arrow-adapter/pkg/otel/traces"
+)
+
+const (
+	dataFormatArrow   = "arrow"
+	receiverTransport = "grpc"
+)
+
+type Consumers interface {
+	Traces() consumer.Traces
+	Metrics() consumer.Metrics
+	Logs() consumer.Logs
+}
+
+type Receiver struct {
+	Consumers
+	arrowpb.UnimplementedEventsServiceServer
+
+	obsrecv       *obsreport.Receiver
+	arrowConsumer *batchEvent.Consumer
+}
+
+// New creates a new Receiver reference.
+func New(
+	id config.ComponentID,
+	cs Consumers,
+	set component.ReceiverCreateSettings,
+) *Receiver {
+	return &Receiver{
+		Consumers: cs,
+		obsrecv: obsreport.NewReceiver(obsreport.ReceiverSettings{
+			ReceiverID:             id,
+			Transport:              receiverTransport,
+			ReceiverCreateSettings: set,
+		}),
+		arrowConsumer: batchEvent.NewConsumer(),
+	}
+}
+
+func (r *Receiver) EventsStream(serverStream arrowpb.EventsService_EventStreamServer) error {
+	ctx := serverStream.Context()
+
+	var traceProducer *traces.OtlpProducer
+
+	if r.Traces() != nil {
+		traceProducer = traces.NewOtlpProducer()
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		resp := &arrowpb.BatchStatus{}
+
+		// Note: How often to send responses?  We can loop here.
+		req, err := serverStream.Recv()
+		if err != nil {
+			return err
+		}
+
+		messages, err := r.arrowConsumer.Consume(req)
+
+		status := &arrowpb.StatusMessage{
+			BatchId: req.GetBatchId(),
+		}
+		if err == nil {
```

Comment:
Should we invert the condition so the main logic isn't in the if? The readability of two if conditions for the same err == nil is a bit odd.